src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,623 @@
+    StartBuildResponse response;
+    try {
+      response = retrier.execute(
+          () ->
+              channel.withChannelBlocking(
+                  channel -> RemoteOutputServiceGrpc.newBlockingStub(channel)
+                      .startBuild(request)));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    // Replace the output path with a symbolic link pointing to the
+    // directory managed by the remote output service.
+    PathFragment outputPath = outputPathPrefix.getRelative(response.getOutputPathSuffix());
+    try {
+      try {
+        originalOutputPath.deleteTree();
+      } catch (FileNotFoundException e) {}
+      originalOutputPath.createSymbolicLink(outputPath);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    stateAcrossBuilds.currentBuildId = buildId;
+    currentRelativeOutputPath = PathFragment.create(relativeOutputPath);
+
+    if (stateAcrossBuilds.previousBuildId == null || !response.hasInitialOutputPathContents()) {
+      // Either Bazel or the remote output service has performed no
+      // build before.
+      return ModifiedFileSet.EVERYTHING_MODIFIED;
+    }
+    InitialOutputPathContents initialContents = response.getInitialOutputPathContents();
+    if (!initialContents.getBuildId().equals(stateAcrossBuilds.previousBuildId.toString())) {
+      // Bazel and the remote output service disagree on the build ID of
+      // the previous build.
+      return ModifiedFileSet.EVERYTHING_MODIFIED;
+    }
+    // Bazel and the remote output service agree on the build ID of the
+    // previous build. Return the set of paths that have been modified.
+    //
+    // TODO: Do these paths need to be relative to the exec root or the
+    // output path? Assume the exec root for now.
+    return ModifiedFileSet.builder()
+        .modifyAll(
+            Iterables.transform(
+                initialContents.getModifiedPathsList(),
+                (p) -> currentRelativeOutputPath.getRelative(p)))
+        .build();
+  }
+
+  @Override
+  public void finalizeBuild(boolean buildSuccessful)
+      throws BuildFailedException, AbruptExitException, InterruptedException {
+    currentBuildSuccessful = buildSuccessful;
+  }
+
+  public void sendFinalizeBuild() {
+    if (stateAcrossBuilds.currentBuildId != null) {
+      FinalizeBuildRequest.Builder builder = FinalizeBuildRequest.newBuilder();
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setBuildSuccessful(currentBuildSuccessful);
+
+      try {
+        channel.withChannelBlocking(
+            channel -> RemoteOutputServiceGrpc.newBlockingStub(channel)
+                .finalizeBuild(builder.build()));
+        stateAcrossBuilds.previousBuildId = stateAcrossBuilds.currentBuildId;
+      } catch (InterruptedException e) {
+        stateAcrossBuilds.previousBuildId = null;
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
```

Comment:
We have seen cases where the bb-clientd daemon dies for whatever reason and Bazel then crashes due to this unchecked exception. This could probably be handled in a less drastic way (by printing the error somehow to tell the user about the problem and exiting with an error code instead of a crash).